### PR TITLE
Add step-by-step tutorial to making institutional configs

### PR DESCRIPTION
**Pull request by jfy133**
This adds a step-by-step tutorial on how to make a institutional config.

### CODE REVIEW

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+subtitle: Walkthrough on what you need to set up an nf-core institutional profile
+---
+
+## Introduction
+
+nf-core offers a centralised place to store Nextflow configuration profiles that work at an _institutional_ level.
+
+What this means is that you can specify common Nextflow pipeline configurations and options that can be shared across all users of your particular institutional cluster.
+
+nf-core offers two level of profile sharing: institutional and pipeline-specific profiles via [nf-core/configs](https://github.com/nf-core/configs)
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:12]: I would rephrase this part, I would not categorise into institutional and pipeline-specific, but global and pipeline-specific. 

```suggestion
nf-core offers two level of institutional profile sharing: global and pipeline-specific profiles via [nf-core/configs](https://github.com/nf-core/configs)
```

Pipeline-specific are still institutional, (except for the viralrecon genomes.config file that is global but pipeline specific.

Comment (MaxUlysse) [reply markdown/usage/tutorials/step_by_step_institutional_profile.md:12]: But I recognise the fact that we actually use institutional most of the time to mean what I called global

Comment (jfy133) [reply markdown/usage/tutorials/step_by_step_institutional_profile.md:12]: What about: `global institutional config` vs . `pipeline-specific institutional config` throughout the text?

Comment (MaxUlysse) [reply markdown/usage/tutorials/step_by_step_institutional_profile.md:12]: Sounds good to me, maybe we should update the nf-core/configs repo for that too?

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+
+executor {
+  queueSize = 8
+  submitRateLimit = '10 sec'
+}
+```
+
+Where a maximum of 10 jobs can be submitted per second
+
+#### <your_container\> scope
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:437]: I'll go for singularity/docker/podman scope here directly.
Or maybe just container scope in the title, and explain that it's either singularity, docker, podman, or even conda...

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+singularity {
+  enabled = true
+  autoMounts = true
+  cacheDir = '/<path>/<to>/<your>/<image_cache>'
+}
+```
+
+Each container engine or software environment may have different options, so be sure to check the Nextflow documentation what options you may have.
+
+#### profiles{} scope
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:474]:
```suggestion
#### profiles scope
```
why adding the `{}`?

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+  config_profile_description = '<your_cluser_name> cluster profile provided by nf-core/configs.'
+  config_profile_contact = '<your_name> (<your_github_handle>)'
+  config_profile_url = 'https://<institutional_url>.com'
+  max_memory = 2.TB
+  max_cpus = 128
+  max_time = 720.h
+}
+```
+
+Finally, if you have a common resource directory for the iGenomes collection of reference genomes, this can can also go in the `params{}` scope.
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:206]:
```suggestion
Finally, if you have a common resource directory for the iGenomes collection of reference genomes, this can can also go in the `params` scope.
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+      max_memory = 256.GB
+      max_cpus = 64
+      max_time = 24.h
+      igenomes_base = '/<path>/<to>/igenomes/'
+    }
+  }
+}
+```
+
+You can see here we have moved the `params{}` block into each of the _profiles_, and updated the `config_profile_description` and `max_*` parameters accordingly.
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:529]:
```suggestion
You can see here we have moved the `params` block into each of the _profiles_, and updated the `config_profile_description` and `max_*` parameters accordingly.
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 6]
```diff
@@ -0,0 +1,617 @@
...
+}
+```
+
+Each container engine or software environment may have different options, so be sure to check the Nextflow documentation what options you may have.
+
+#### profiles{} scope
+
+In some cases, you may want to define multiple contexts that have different specifications - For example, your institution may have has two distinct clusters or set of nodes with slightly different specifications, while still sharing the same storage.
+
+One concise way of specifying these is via the `profiles{}` scope. These essentially act as nested configs, where you can extend or modify the base parameters set in the main config.
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:478]:
```suggestion
One concise way of specifying these is via the `profiles` scope. These essentially act as nested configs, where you can extend or modify the base parameters set in the main config.
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 7]
```diff
@@ -0,0 +1,617 @@
...
+
+To test your config, pick your pipeline of choice, and run the given pipeline's integrated mini-test profile but using your config. For example:
+
+```bash
+nextflow run nf-core/eager \
+-profile <your_cluster_name>,test_tsv \
+--custom_config_base 'https://raw.githubusercontent.com/<your_github_user>/configs/<your_branch>'
+```
+
+If you also wish to test a profile from your `profiles{}` scope, include that in the `-profile` flag. For example:
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:553]:
```suggestion
If you also wish to test a profile from your `profiles` scope, include that in the `-profile` flag. For example:
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 8]
```diff
@@ -0,0 +1,617 @@
...
+
+```nextflow
+debug {
+  cleanup = false
+}
+```
+
+which allows you to override the default `cleanup = true` behaviour, if you've set this as so.
+
+Alternatively, if your cluster utilises a scratch space to store intermediate files, this can be specified a the `process{}` scope of the profile with the [`scratch`](https://www.nextflow.io/docs/latest/process.html#scratch) directive. For example:
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:610]:
```suggestion
Alternatively, if your cluster utilises a scratch space to store intermediate files, this can be specified a the `process` scope of the profile with the [`scratch`](https://www.nextflow.io/docs/latest/process.html#scratch) directive. For example:
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+
+- Describe commonly useful information that is worth gathering _before_ writing such a profile
+- Go through each component of the profile to describe how this should be written
+- Show how to test such a profile before submitting it to nf-core
+
+## Preparation
+
+We will first go through a list of questions to ask and what information to gather prior writing the profile. Not all questions will be relevant to your cluster, however it should cover a range of common set of cluster specifications.
+
+## Does an institutional profile already exist
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:27]:
```suggestion
### Does an institutional profile already exist
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 10]
```diff
@@ -0,0 +1,617 @@
...
+
+First you should check nf-core/configs an institutional config exists! If you see any issues with it, or it's not working for you - contact the person who originally made it.
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:29]:
```suggestion
First you should check nf-core/configs an institutional profile exists! If you see any issues with it, or it's not working for you - contact the person who originally made it.
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 11]
```diff
@@ -0,0 +1,617 @@
...
+
+If no config already exists, continue with this tutorial!
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:31]:
```suggestion
If no profile already exists, continue with this tutorial!
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 12]
```diff
@@ -0,0 +1,617 @@
...
+
+### At what level should the institutional profile be designed
+
+Various institutions have different structures in the way clusters are accessed by different groups.
+
+This is an important consideration to make before starting writing a profile because you need to select a useful and descriptive name.
+
+Sometimes one 'institution' will have multiple 'clusters' with the different names, which may share a common storage space. In this case it might make sense to have one nf-core configuration file for the whole institution, but with multiple _internal_ profiles representing each cluster.
+
+Alternatively, within one 'institution', you may have groups or departments with entirely different and separated clusters with dedicated IT teams. In this case it may make sense to have separate configuration files without a common institutional 'umbrella' name.
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:41]:
```suggestion
Sometimes one institution will have multiple clusters with the different names, which may share a common storage space. In this case it might make sense to have one nf-core configuration file for the whole institution, but with multiple _internal_ profiles representing each cluster.

Alternatively, within one institution, you may have groups or departments with entirely different and separated clusters with dedicated IT teams. In this case it may make sense to have separate configuration files without a common institutional 'umbrella' name.
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 13]
```diff
@@ -0,0 +1,617 @@
...
+
+Nextflow has integrated a range of scheduling systems to simplify running Nextflow pipelines on a range of HPCs. What this means is that it will write and sumbit submission scripts to your given scheduler or 'executor' (in Nextflow language) for you.
+
+You can see the range of support schedulers on the [Nextflow documentation](https://www.Nextflow.io/docs/latest/executor.html). Check if your scheduler is there, and how it would be written (typically the name in all lower case).
+
+You should note down if there are any special options that your cluster's scheduler requires. This maybe special ways in which memory requirements are specified, e.g.: memory per core, types of parallelisation library is required etc., the maximum number of parallel jobs that are allowed.
+
+<details markdown='1'><summary> SGE tips</summary>
+
+Some SGE configurations require the use of the `-l h_vmem=` flag _in addition_ to `-l virtual_free=` - the latter of which is the default used by default in Nextflow.
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:61]: `the default used by default`?
```suggestion
Some SGE configurations require the use of the `-l h_vmem=` flag _in addition_ to `-l virtual_free=` - the latter of which is used by default in Nextflow.
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 14]
```diff
@@ -0,0 +1,617 @@
...
+sinfo
+```
+
+You should record the names of each `PARTITION` and the corresponding `TIMELIMIT` column. Note there may be additional resource limitations for each queue, so check your institutional cluster's documentation for any other limits.
+
+</details>
+
+## Are there any common resource or data locations on the cluster
+
+For institutions that work on common topics (e.g. genomics), you might have a centralised local copy of the Illumina `iGenome` set of reference genomes. If it exists, you can note down the paths to this directory.
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:132]:
```suggestion
For institutions that work on common topics (e.g. genomics), you might have a centralised local copy of the AWS `iGenomes` set of reference genomes. If it exists, you can note down the paths to this directory.
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 15]
```diff
@@ -0,0 +1,617 @@
...
+
+Furthermore, some institutions will specify a common place to store all container images (e.g. for Docker, Singularity etc) in a cache directory. This helps preventing multiple users having their own copies of the same pipeline container image. You should find out if, and where this location exists and note down the path.
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:134]: I don't think there is a cache dir option for Docker

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 16]
```diff
@@ -0,0 +1,617 @@
...
+
+```bash
+sinfo
+```
+
+You should record the names of each `PARTITION` and the corresponding `TIMELIMIT` column. Note there may be additional resource limitations for each queue, so check your institutional cluster's documentation for any other limits.
+
+</details>
+
+## Are there any common resource or data locations on the cluster
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:130]:
```suggestion
### Are there any common resource or data locations on the cluster
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 17]
```diff
@@ -0,0 +1,617 @@
...
+
+For institutions that work on common topics (e.g. genomics), you might have a centralised local copy of the Illumina `iGenome` set of reference genomes. If it exists, you can note down the paths to this directory.
+
+Furthermore, some institutions will specify a common place to store all container images (e.g. for Docker, Singularity etc) in a cache directory. This helps preventing multiple users having their own copies of the same pipeline container image. You should find out if, and where this location exists and note down the path.
+
+## Should jobs be run using a local `$TMPDIR`
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:136]:
```suggestion
### Should jobs be run using a local `$TMPDIR`
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 18]
```diff
@@ -0,0 +1,617 @@
...
+
+Some clusters require the use of a temporary _per-node_ `scratch` directory, the location of which is specified in an environmental variable `$TMPDIR`. Check if this is the case and this is set up in `$TMPDIR`
+
+## Preparation checklist
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:140]:
```suggestion
### Preparation checklist
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 19]
```diff
@@ -0,0 +1,617 @@
...
+To summarise, before you start writing the profile have you:
+
+- Checked you know what institutional level you will make the profile at
+- Checked permission to publicly post the profile
+- Checked which scheduler the cluster uses
+- Checked which container engine you will specify
+- Checked which maximum computational resource limits exist
+- Checked for any queue/partition names and specifications
+- Checked for locations of any common resources or cache directories
+- Checked if jobs on the cluster require use of a per-node `scratch`
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:151]:
```suggestion
To summarise, before you start writing the profile have you checked:

- At what institutional level you will make the profile
- The permission to publicly post the profile
- Which scheduler the cluster uses
- Which container engine you will specify
- Which maximum computational resource limits exist
- Queue/partition names and specifications
- Locations of any common resources or cache directories
- If jobs on the cluster require use of a per-node `scratch`
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 20]
```diff
@@ -0,0 +1,617 @@
...
+
+## Creating the profile
+
+Once you have all the information ready, we can begin writing the institutional profile! The tutorial will now go through each relevant Nextflow 'scope' block and explain what and when to put information in.
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:155]:
```suggestion
Once you have all the information ready, we can begin writing the institutional profile! The walkthrough will now go through each relevant Nextflow 'scope' block and explain what and when to put information in.
```

File: markdown/usage/tutorials/step_by_step_institutional_profile.md [hunk 21]
```diff
@@ -0,0 +1,617 @@
...
+
+Before you start consider what your institutional profile should be called. The names should be as unique and recognisable as possible - however abbreviations or acronyms are fine.
+
+For example, the name of the cluster can be used (e.g. [`phoenix`](https://github.com/nf-core/configs/blob/master/conf/phoenix.config) or [binac](https://github.com/nf-core/configs/blob/master/conf/binac.config)). In some cases an acronym of the cluster or the institute can be selected (e.g. [`SHH`](https://github.com/nf-core/configs/blob/master/conf/shh.config) or [`EVA`](https://github.com/nf-core/configs/blob/master/conf/eva.config)).
+
+Next, make sure to make your own _fork_ of [nf-core/configs](https://github.com/nf-core/configs), and make a new branch to work on. It's normally best to call the branch the name of your proposed cluster name. Clone this to your machine, and open the folder.
```
Comment (MaxUlysse) [inline markdown/usage/tutorials/step_by_step_institutional_profile.md:161]: It's not necessary to make a branch, usually people would work on their own fork of nf-core/configs once, and that's all.
In any other case, I'd recommend making a branch, in this case I think it's less important.
But as it's a good practice, I'm all for it.

**Comment by MaxUlysse** [comment]
Looks very nice and quite complete.
I'd just add a section about hostnames: https://github.com/nf-core/configs/blob/master/nfcore_custom.config#L55-L73
And maybe some tips on how it can be use: https://github.com/nf-core/configs/blob/master/conf/uppmax.config#L30-L42

File: markdown/usage/tutorials/step_by_step_institutional_profile.md (hunk 15 above)
Comment (jfy133) [reply markdown/usage/tutorials/step_by_step_institutional_profile.md:134]: Huh, just checked the nextflow docs and you're right. That's wierd.

File: markdown/usage/tutorials/step_by_step_institutional_profile.md (hunk 21 above)
Comment (jfy133) [reply markdown/usage/tutorials/step_by_step_institutional_profile.md:161]: I also originally did that until recently when I got stuck doing loads of debugging and messed up master, so then couldn't pull in the latest nf-core/configs master. So I think while it's less likely, I feel it's good practise anyway (like you said) 😬

File: markdown/usage/tutorials/step_by_step_institutional_profile.md (hunk 1 above)
Comment (jfy133) [reply markdown/usage/tutorials/step_by_step_institutional_profile.md]: Probably should! Will do after this is in

**Review by maxulysse — APPROVED** [review]
A bit too many emojis for my taste, but LGTM
